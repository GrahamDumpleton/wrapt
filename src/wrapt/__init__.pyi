import sys

if sys.version_info >= (3, 10):
    from inspect import FullArgSpec
    from types import ModuleType, TracebackType
    from typing import (
        Any,
        Callable,
        Concatenate,
        Generic,
        ParamSpec,
        Protocol,
        TypeVar,
        overload,
    )

    P = ParamSpec("P")
    R = TypeVar("R", covariant=True)

    T = TypeVar("T", bound=Any)

    class Boolean(Protocol):
        def __bool__(self) -> bool: ...

    # ObjectProxy

    class ObjectProxy(Generic[T]):
        __wrapped__: T
        def __init__(self, wrapped: T) -> None: ...

    # CallableObjectProxy

    class CallableObjectProxy(ObjectProxy[T]):
        def __call__(self, *args: Any, **kwargs: Any) -> Any: ...

    # PartialCallableObjectProxy

    class PartialCallableObjectProxy:
        def __init__(
            self, func: Callable[..., Any], *args: Any, **kwargs: Any
        ) -> None: ...
        def __call__(self, *args: Any, **kwargs: Any) -> Any: ...

    def partial(
        func: Callable[..., Any], /, *args: Any, **kwargs: Any
    ) -> Callable[..., Any]: ...

    # WeakFunctionProxy

    class WeakFunctionProxy:
        def __init__(
            self,
            wrapped: Callable[..., Any],
            callback: Callable[..., Any] | None = None,
        ) -> None: ...
        def __call__(self, *args: Any, **kwargs: Any) -> Any: ...

    # FunctionWrapper

    WrappedFunction = Callable[P, R]

    GenericCallableWrapperFunction = Callable[
        [WrappedFunction[P, R], Any, tuple[Any, ...], dict[str, Any]], R
    ]

    ClassMethodWrapperFunction = Callable[
        [type[Any], WrappedFunction[P, R], Any, tuple[Any, ...], dict[str, Any]], R
    ]

    InstanceMethodWrapperFunction = Callable[
        [Any, WrappedFunction[P, R], Any, tuple[Any, ...], dict[str, Any]], R
    ]

    WrapperFunction = (
        GenericCallableWrapperFunction[P, R]
        | ClassMethodWrapperFunction[P, R]
        | InstanceMethodWrapperFunction[P, R]
    )

    class _FunctionWrapperBase(ObjectProxy[WrappedFunction[P, R]]):
        _self_instance: Any
        _self_wrapper: WrapperFunction[P, R]
        _self_enabled: bool | Boolean | Callable[[], bool] | None
        _self_binding: str
        _self_parent: Any
        _self_owner: Any

    class BoundFunctionWrapper(_FunctionWrapperBase[P, R]):
        def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R: ...
        def __get__(
            self, instance: Any, owner: type[Any] | None = None
        ) -> "BoundFunctionWrapper[P, R]": ...

    class FunctionWrapper(_FunctionWrapperBase[P, R]):
        def __init__(
            self,
            wrapped: WrappedFunction[P, R],
            wrapper: WrapperFunction[P, R],
            enabled: bool | Boolean | Callable[[], bool] | None = None,
        ) -> None: ...
        def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R: ...
        def __get__(
            self, instance: Any, owner: type[Any] | None = None
        ) -> BoundFunctionWrapper[P, R]: ...

    # AdapterFactory/adapter_factory()

    class AdapterFactory(Protocol):
        def __call__(
            self, wrapped: Callable[..., Any]
        ) -> str | FullArgSpec | Callable[..., Any]: ...

    def adapter_factory(wrapped: Callable[..., Any]) -> AdapterFactory: ...

    # decorator()

    class Descriptor(Protocol):
        def __get__(self, instance: Any, owner: type[Any] | None = None) -> Any: ...

    class FunctionDecorator(Generic[P, R]):
        def __call__(
            self,
            callable: (
                Callable[P, R]
                | Callable[Concatenate[type[T], P], R]
                # | Callable[Concatenate[Any, P], R] # Don't use, breaks stuff.
                | Callable[[type[T]], R]
                | Descriptor
            ),
        ) -> FunctionWrapper[P, R]: ...

    class PartialFunctionDecorator:
        @overload
        def __call__(
            self, wrapper: GenericCallableWrapperFunction[P, R], /
        ) -> FunctionDecorator[P, R]: ...
        @overload
        def __call__(
            self, wrapper: ClassMethodWrapperFunction[P, R], /
        ) -> FunctionDecorator[P, R]: ...
        @overload
        def __call__(
            self, wrapper: InstanceMethodWrapperFunction[P, R], /
        ) -> FunctionDecorator[P, R]: ...

    # ... Decorator applied to class type.

    @overload
    def decorator(wrapper: type[T], /) -> FunctionDecorator[Any, Any]: ...

    # ... Decorator applied to function or method.

    @overload
    def decorator(
        wrapper: GenericCallableWrapperFunction[P, R], /
    ) -> FunctionDecorator[P, R]: ...
    @overload
    def decorator(
        wrapper: ClassMethodWrapperFunction[P, R], /
    ) -> FunctionDecorator[P, R]: ...
    @overload
    def decorator(
        wrapper: InstanceMethodWrapperFunction[P, R], /
    ) -> FunctionDecorator[P, R]: ...

    # ... Positional arguments.

    @overload
    def decorator(
        *,
        enabled: bool | Boolean | Callable[[], bool] | None = None,
        adapter: str | FullArgSpec | AdapterFactory | Callable[..., Any] | None = None,
        proxy: type[FunctionWrapper[Any, Any]] | None = None,
    ) -> PartialFunctionDecorator: ...

    # function_wrapper()

    @overload
    def function_wrapper(wrapper: type[Any]) -> FunctionDecorator[Any, Any]: ...
    @overload
    def function_wrapper(
        wrapper: GenericCallableWrapperFunction[P, R],
    ) -> FunctionDecorator[P, R]: ...
    @overload
    def function_wrapper(
        wrapper: ClassMethodWrapperFunction[P, R],
    ) -> FunctionDecorator[P, R]: ...
    @overload
    def function_wrapper(
        wrapper: InstanceMethodWrapperFunction[P, R],
    ) -> FunctionDecorator[P, R]: ...
    # @overload
    # def function_wrapper(wrapper: Any) -> FunctionDecorator[Any, Any]: ... # Don't use, breaks stuff.

    # wrap_function_wrapper()

    def wrap_function_wrapper(
        target: ModuleType | type[Any] | Any | str,
        name: str,
        wrapper: WrapperFunction[P, R],
    ) -> FunctionWrapper[P, R]: ...

    # patch_function_wrapper()

    class WrapperDecorator:
        def __call__(self, wrapper: WrapperFunction[P, R]) -> FunctionWrapper[P, R]: ...

    def patch_function_wrapper(
        target: ModuleType | type[Any] | Any | str,
        name: str,
        enabled: bool | Boolean | Callable[[], bool] | None = None,
    ) -> WrapperDecorator: ...

    # transient_function_wrapper()

    class TransientDecorator:
        def __call__(
            self, wrapper: WrapperFunction[P, R]
        ) -> FunctionDecorator[P, R]: ...

    def transient_function_wrapper(
        target: ModuleType | type[Any] | Any | str, name: str
    ) -> TransientDecorator: ...

    # resolve_path()

    def resolve_path(
        target: ModuleType | type[Any] | Any | str, name: str
    ) -> tuple[ModuleType | type[Any] | Any, str, Callable[..., Any]]: ...

    # apply_patch()

    def apply_patch(
        parent: ModuleType | type[Any] | Any,
        attribute: str,
        replacement: Any,
    ) -> None: ...

    # wrap_object()

    WrapperFactory = Callable[
        [Callable[..., Any], tuple[Any, ...], dict[str, Any]], type[ObjectProxy[Any]]
    ]

    def wrap_object(
        target: ModuleType | type[Any] | Any | str,
        name: str,
        factory: WrapperFactory | type[ObjectProxy[Any]],
        args: tuple[Any, ...],
        kwargs: dict[str, Any],
    ) -> Any: ...

    # wrap_object_attribute()

    def wrap_object_attribute(
        target: ModuleType | type[Any] | Any | str,
        name: str,
        factory: WrapperFactory | type[ObjectProxy[Any]],
        args: tuple[Any, ...] = (),
        kwargs: dict[str, Any] = {},
    ) -> Any: ...

    # register_post_import_hook()

    def register_post_import_hook(
        hook: Callable[[ModuleType], Any] | str, name: str
    ) -> None: ...

    # discover_post_import_hooks()

    def discover_post_import_hooks(group: str) -> None: ...

    # notify_module_loaded()

    def notify_module_loaded(module: ModuleType) -> None: ...

    # when_imported()

    class ImportHookDecorator:
        def __call__(self, hook: Callable[[ModuleType], Any]) -> Callable[..., Any]: ...

    def when_imported(name: str) -> ImportHookDecorator: ...

    # synchronized()

    class SynchronizedObject:
        def __call__(self, wrapped: Callable[P, R]) -> Callable[P, R]: ...
        def __enter__(self) -> Any: ...
        def __exit__(
            self,
            exc_type: type[BaseException] | None,
            exc_value: BaseException | None,
            traceback: TracebackType | None,
        ) -> bool | None: ...

    @overload
    def synchronized(wrapped: Callable[P, R]) -> Callable[P, R]: ...
    @overload
    def synchronized(wrapped: Any) -> SynchronizedObject: ...
