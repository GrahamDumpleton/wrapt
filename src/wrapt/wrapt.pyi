from typing import Any, Generic, TypeVar, Callable, Optional

F = TypeVar('F', bound=Callable[..., Any])
A = TypeVar('A', bound=Callable[..., Any])
T = TypeVar("T", bound=Any)

def decorator(wrapper: F, enabled: Optional[bool] = None, adapter: Optional[A] = None) -> F: ...

class ObjectProxy(Generic[T]):
    __wrapped__ : T

    def __init__(self, wrapped: T):
        ...
