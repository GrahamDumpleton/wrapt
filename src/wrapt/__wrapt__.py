"""This module is used to switch between C and Python implementations of the
wrappers.
"""

import os

from .wrappers import (
    BoundFunctionWrapper,
    CallableObjectProxy,
    FunctionWrapper,
    ObjectProxy,
    PartialCallableObjectProxy,
    _FunctionWrapperBase,
)

# Import Python fallbacks.


# Try to use C extensions if not disabled.

_use_extensions = not os.environ.get("WRAPT_DISABLE_EXTENSIONS")

if _use_extensions:
    try:
        from ._wrappers import (  # type: ignore[no-redef,import-not-found]
            BoundFunctionWrapper,
            CallableObjectProxy,
            FunctionWrapper,
            ObjectProxy,
            PartialCallableObjectProxy,
            _FunctionWrapperBase,
        )
    except ImportError:
        # C extensions not available, using Python implementations
        pass

# Provide an alias for partial().


def partial(*args, **kwargs):
    """Create a callable object proxy with partial application of the given
    arguments and keywords. This behaves the same as `functools.partial`, but
    implemented using the `ObjectProxy` class to provide better support for
    introspection.
    """
    return PartialCallableObjectProxy(*args, **kwargs)
