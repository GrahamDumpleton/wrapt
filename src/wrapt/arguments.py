"""The inspect.formatargspec() function was dropped in Python 3.11 but we need
it for when constructing signature changing decorators based on result of
inspect.getfullargspec(). The code here implements inspect.formatargspec() based
on Parameter and Signature from inspect module, which were added in Python 3.6.
Thanks to Cyril Jouve for the implementation.
"""

from typing import Any, Callable, List, Mapping, Optional, Sequence, Tuple

try:
    from inspect import Parameter, Signature
except ImportError:
    from inspect import formatargspec  # type: ignore[attr-defined]
else:

    def formatargspec(
        args: List[str],
        varargs: Optional[str] = None,
        varkw: Optional[str] = None,
        defaults: Optional[Tuple[Any, ...]] = None,
        kwonlyargs: Optional[Sequence[str]] = None,
        kwonlydefaults: Optional[Mapping[str, Any]] = None,
        annotations: Mapping[str, Any] = {},
        formatarg: Callable[[str], str] = str,
        formatvarargs: Callable[[str], str] = lambda name: "*" + name,
        formatvarkw: Callable[[str], str] = lambda name: "**" + name,
        formatvalue: Callable[[Any], str] = lambda value: "=" + repr(value),
        formatreturns: Callable[[Any], str] = lambda text: " -> " + text,
        formatannotation: Callable[[Any], str] = lambda annot: " -> " + repr(annot),
    ) -> str:
        if kwonlyargs is None:
            kwonlyargs = ()
        if kwonlydefaults is None:
            kwonlydefaults = {}
        ndefaults = len(defaults) if defaults else 0
        parameters = [
            Parameter(
                arg,
                Parameter.POSITIONAL_OR_KEYWORD,
                default=defaults[i] if defaults and i >= 0 else Parameter.empty,
                annotation=annotations.get(arg, Parameter.empty),
            )
            for i, arg in enumerate(args, ndefaults - len(args))
        ]
        if varargs:
            parameters.append(Parameter(varargs, Parameter.VAR_POSITIONAL))
        parameters.extend(
            Parameter(
                kwonlyarg,
                Parameter.KEYWORD_ONLY,
                default=kwonlydefaults.get(kwonlyarg, Parameter.empty),
                annotation=annotations.get(kwonlyarg, Parameter.empty),
            )
            for kwonlyarg in kwonlyargs
        )
        if varkw:
            parameters.append(Parameter(varkw, Parameter.VAR_KEYWORD))
        return_annotation = annotations.get("return", Signature.empty)
        return str(Signature(parameters, return_annotation=return_annotation))
