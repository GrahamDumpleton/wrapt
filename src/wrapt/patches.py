import inspect
import sys

from .__wrapt__ import FunctionWrapper

# Helper functions for applying wrappers to existing functions.


def resolve_path(target, name):
    """
    Resolves the dotted path supplied as `name` to an attribute on a target
    object. The `target` can be a module, class, or instance of a class. If the
    `target` argument is a string, it is assumed to be the name of a module,
    which will be imported if necessary and then used as the target object.
    Returns a tuple containing the parent object holding the attribute lookup
    resolved to, the attribute name (path prefix removed if present), and the
    original attribute value.
    """

    if isinstance(target, str):
        __import__(target)
        target = sys.modules[target]

    parent = target

    path = name.split(".")
    attribute = path[0]

    # We can't just always use getattr() because in doing
    # that on a class it will cause binding to occur which
    # will complicate things later and cause some things not
    # to work. For the case of a class we therefore access
    # the __dict__ directly. To cope though with the wrong
    # class being given to us, or a method being moved into
    # a base class, we need to walk the class hierarchy to
    # work out exactly which __dict__ the method was defined
    # in, as accessing it from __dict__ will fail if it was
    # not actually on the class given. Fallback to using
    # getattr() if we can't find it. If it truly doesn't
    # exist, then that will fail.

    def lookup_attribute(parent, attribute):
        if inspect.isclass(parent):
            for cls in inspect.getmro(parent):
                if attribute in vars(cls):
                    return vars(cls)[attribute]
            else:
                return getattr(parent, attribute)
        else:
            return getattr(parent, attribute)

    original = lookup_attribute(parent, attribute)

    for attribute in path[1:]:
        parent = original
        original = lookup_attribute(parent, attribute)

    return (parent, attribute, original)


def apply_patch(parent, attribute, replacement):
    """
    Convenience function for applying a patch to an attribute. Currently this
    maps to the standard setattr() function, but in the future may be extended
    to support more complex patching strategies.
    """

    setattr(parent, attribute, replacement)


def wrap_object(target, name, factory, args=(), kwargs={}):
    """
    Wraps an object which is the attribute of a target object with a wrapper
    object created by the `factory` function. The `target` can be a module,
    class, or instance of a class. In the special case of `target` being a
    string, it is assumed to be the name of a module, with the module being
    imported if necessary and then used as the target object. The `name` is a
    string representing the dotted path to the attribute. The `factory` function
    should accept the original object and may accept additional positional and
    keyword arguments which will be set by unpacking input arguments using
    `*args` and `**kwargs` calling conventions. The factory function should
    return a new object that will replace the original object.
    """

    (parent, attribute, original) = resolve_path(target, name)
    wrapper = factory(original, *args, **kwargs)
    apply_patch(parent, attribute, wrapper)

    return wrapper


# Function for applying a proxy object to an attribute of a class
# instance. The wrapper works by defining an attribute of the same name
# on the class which is a descriptor and which intercepts access to the
# instance attribute. Note that this cannot be used on attributes which
# are themselves defined by a property object.


class AttributeWrapper:

    def __init__(self, attribute, factory, args, kwargs):
        self.attribute = attribute
        self.factory = factory
        self.args = args
        self.kwargs = kwargs

    def __get__(self, instance, owner):
        value = instance.__dict__[self.attribute]
        return self.factory(value, *self.args, **self.kwargs)

    def __set__(self, instance, value):
        instance.__dict__[self.attribute] = value

    def __delete__(self, instance):
        del instance.__dict__[self.attribute]


def wrap_object_attribute(module, name, factory, args=(), kwargs={}):
    """
    Wraps an object which is the attribute of a class instance with a wrapper
    object created by the `factory` function. It does this by patching the
    class, not the instance, with a descriptor that intercepts access to the
    instance attribute. The `module` can be a module, class, or instance of a
    class. In the special case of `module` being a string, it is assumed to be
    the name of a module, with the module being imported if necessary and then
    used as the target object. The `name` is a string representing the dotted
    path to the attribute. The `factory` function should accept the original
    object and may accept additional positional and keyword arguments which will
    be set by unpacking input arguments using `*args` and `**kwargs` calling
    conventions. The factory function should return a new object that will
    replace the original object.
    """

    path, attribute = name.rsplit(".", 1)
    parent = resolve_path(module, path)[2]
    wrapper = AttributeWrapper(attribute, factory, args, kwargs)
    apply_patch(parent, attribute, wrapper)
    return wrapper


# Functions for creating a simple decorator using a FunctionWrapper,
# plus short cut functions for applying wrappers to functions. These are
# for use when doing monkey patching. For a more featured way of
# creating decorators see the decorator decorator instead.


def function_wrapper(wrapper):
    """
    Creates a decorator for wrapping a function with a `wrapper` function.
    The decorator which is returned may also be applied to any other callable
    objects such as lambda functions, methods, classmethods, and staticmethods,
    or objects which implement the `__call__()` method. The `wrapper` function
    should accept the `wrapped` function, `instance`, `args`, and `kwargs`,
    arguments and return the result of calling the wrapped function or some
    other appropriate value.
    """

    def _wrapper(wrapped, instance, args, kwargs):
        target_wrapped = args[0]
        if instance is None:
            target_wrapper = wrapper
        elif inspect.isclass(instance):
            target_wrapper = wrapper.__get__(None, instance)
        else:
            target_wrapper = wrapper.__get__(instance, type(instance))
        return FunctionWrapper(target_wrapped, target_wrapper)

    return FunctionWrapper(wrapper, _wrapper)


def wrap_function_wrapper(target, name, wrapper):
    """
    Wraps a function which is the attribute of a target object with a `wrapper`
    function. The `target` can be a module, class, or instance of a class. In
    the special case of `target` being a string, it is assumed to be the name
    of a module, with the module being imported if necessary. The `name` is a
    string representing the dotted path to the attribute. The `wrapper` function
    should accept the `wrapped` function, `instance`, `args`, and `kwargs`
    arguments, and would return the result of calling the wrapped attribute or
    some other appropriate value.
    """

    return wrap_object(target, name, FunctionWrapper, (wrapper,))


def patch_function_wrapper(target, name, enabled=None):
    """
    Creates a decorator which can be applied to a wrapper function, where the
    wrapper function will be used to wrap a function which is the attribute of
    a target object. The `target` can be a module, class, or instance of a class.
    In the special case of `target` being a string, it is assumed to be the name
    of a module, with the module being imported if necessary. The `name` is a
    string representing the dotted path to the attribute. The `enabled`
    argument can be a boolean or a callable that returns a boolean. When a
    callable is provided, it will be called each time the wrapper is invoked to
    determine if the wrapper function should be executed or whether the wrapped
    function should be called directly. If `enabled` is not provided, the
    wrapper is enabled by default.
    """

    def _wrapper(wrapper):
        return wrap_object(target, name, FunctionWrapper, (wrapper, enabled))

    return _wrapper


def transient_function_wrapper(target, name):
    """Creates a decorator that patches a target function with a wrapper
    function, but only for the duration of the call that the decorator was
    applied to. The `target` can be a module, class, or instance of a class.
    In the special case of `target` being a string, it is assumed to be the name
    of a module, with the module being imported if necessary. The `name` is a
    string representing the dotted path to the attribute.
    """

    def _decorator(wrapper):
        def _wrapper(wrapped, instance, args, kwargs):
            target_wrapped = args[0]
            if instance is None:
                target_wrapper = wrapper
            elif inspect.isclass(instance):
                target_wrapper = wrapper.__get__(None, instance)
            else:
                target_wrapper = wrapper.__get__(instance, type(instance))

            def _execute(wrapped, instance, args, kwargs):
                (parent, attribute, original) = resolve_path(target, name)
                replacement = FunctionWrapper(original, target_wrapper)
                setattr(parent, attribute, replacement)
                try:
                    return wrapped(*args, **kwargs)
                finally:
                    setattr(parent, attribute, original)

            return FunctionWrapper(target_wrapped, _execute)

        return FunctionWrapper(wrapper, _wrapper)

    return _decorator
