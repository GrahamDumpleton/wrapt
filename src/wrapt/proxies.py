"""Variants of ObjectProxy for different use cases."""

from .__wrapt__ import BaseObjectProxy
from .decorators import synchronized

# Define ObjectProxy which for compatibility adds `__iter__()` support which
# has been removed from `BaseObjectProxy`.


class ObjectProxy(BaseObjectProxy):
    """A generic object proxy which forwards special methods as needed.
    For backwards compatibility this class adds support for `__iter__()`. If
    you don't need backward compatibility for `__iter__()` support then it is
    preferable to use `BaseObjectProxy` directly. If you want automatic
    support for special dunder methods for callables, iterators, and async,
    then use `AutoObjectProxy`."""

    @property
    def __object_proxy__(self):
        return ObjectProxy

    def __new__(cls, *args, **kwargs):
        return super().__new__(cls)

    def __iter__(self):
        return iter(self.__wrapped__)


# Define variant of ObjectProxy which can automatically adjust to the wrapped
# object and add special dunder methods.


def __wrapper_call__(self, *args, **kwargs):
    return self.__wrapped__(*args, **kwargs)


def __wrapper_iter__(self):
    return iter(self.__wrapped__)


def __wrapper_next__(self):
    return self.__wrapped__.__next__()


def __wrapper_aiter__(self):
    return self.__wrapped__.__aiter__()


async def __wrapper_anext__(self):
    return await self.__wrapped__.__anext__()


def __wrapper_length_hint__(self):
    return self.__wrapped__.__length_hint__()


def __wrapper_await__(self):
    return (yield from self.__wrapped__.__await__())


def __wrapper_get__(self, instance, owner):
    return self.__wrapped__.__get__(instance, owner)


def __wrapper_set__(self, instance, value):
    return self.__wrapped__.__set__(instance, value)


def __wrapper_delete__(self, instance):
    return self.__wrapped__.__delete__(instance)


def __wrapper_set_name__(self, owner, name):
    return self.__wrapped__.__set_name__(owner, name)


class AutoObjectProxy(BaseObjectProxy):
    """An object proxy which can automatically adjust to the wrapped object
    and add special dunder methods as needed. Note that this creates a new
    class for each instance, so it has much higher memory overhead than using
    `BaseObjectProxy` directly. If you know what special dunder methods you need
    then it is preferable to use `BaseObjectProxy` directly and add them to a
    subclass as needed. If you only need `__iter__()` support for backwards
    compatibility then use `ObjectProxy` instead.
    """

    def __new__(cls, wrapped):
        """Injects special dunder methods into a dynamically created subclass
        as needed based on the wrapped object.
        """

        namespace = {}

        wrapped_attrs = dir(wrapped)
        class_attrs = set(dir(cls))

        if callable(wrapped) and "__call__" not in class_attrs:
            namespace["__call__"] = __wrapper_call__

        if "__iter__" in wrapped_attrs and "__iter__" not in class_attrs:
            namespace["__iter__"] = __wrapper_iter__

        if "__next__" in wrapped_attrs and "__next__" not in class_attrs:
            namespace["__next__"] = __wrapper_next__

        if "__aiter__" in wrapped_attrs and "__aiter__" not in class_attrs:
            namespace["__aiter__"] = __wrapper_aiter__

        if "__anext__" in wrapped_attrs and "__anext__" not in class_attrs:
            namespace["__anext__"] = __wrapper_anext__

        if "__length_hint__" in wrapped_attrs and "__length_hint__" not in class_attrs:
            namespace["__length_hint__"] = __wrapper_length_hint__

        # Note that not providing compatibility with generator-based coroutines
        # (PEP 342) here as they are removed in Python 3.11+ and were deprecated
        # in 3.8.

        if "__await__" in wrapped_attrs and "__await__" not in class_attrs:
            namespace["__await__"] = __wrapper_await__

        if "__get__" in wrapped_attrs and "__get__" not in class_attrs:
            namespace["__get__"] = __wrapper_get__

        if "__set__" in wrapped_attrs and "__set__" not in class_attrs:
            namespace["__set__"] = __wrapper_set__

        if "__delete__" in wrapped_attrs and "__delete__" not in class_attrs:
            namespace["__delete__"] = __wrapper_delete__

        if "__set_name__" in wrapped_attrs and "__set_name__" not in class_attrs:
            namespace["__set_name__"] = __wrapper_set_name__

        name = cls.__name__

        if cls is AutoObjectProxy:
            name = BaseObjectProxy.__name__

        return super().__new__(type(name, (cls,), namespace))

    def __wrapped_setattr_fixups__(self):
        """Adjusts special dunder methods on the class as needed based on the
        wrapped object, when `__wrapped__` is changed.
        """

        cls = type(self)
        class_attrs = set(dir(cls))

        if callable(self.__wrapped__):
            if "__call__" not in class_attrs:
                cls.__call__ = __wrapper_call__
        elif getattr(cls, "__call__", None) is __wrapper_call__:
            delattr(cls, "__call__")

        if hasattr(self.__wrapped__, "__iter__"):
            if "__iter__" not in class_attrs:
                cls.__iter__ = __wrapper_iter__
        elif getattr(cls, "__iter__", None) is __wrapper_iter__:
            delattr(cls, "__iter__")

        if hasattr(self.__wrapped__, "__next__"):
            if "__next__" not in class_attrs:
                cls.__next__ = __wrapper_next__
        elif getattr(cls, "__next__", None) is __wrapper_next__:
            delattr(cls, "__next__")

        if hasattr(self.__wrapped__, "__aiter__"):
            if "__aiter__" not in class_attrs:
                cls.__aiter__ = __wrapper_aiter__
        elif getattr(cls, "__aiter__", None) is __wrapper_aiter__:
            delattr(cls, "__aiter__")

        if hasattr(self.__wrapped__, "__anext__"):
            if "__anext__" not in class_attrs:
                cls.__anext__ = __wrapper_anext__
        elif getattr(cls, "__anext__", None) is __wrapper_anext__:
            delattr(cls, "__anext__")

        if hasattr(self.__wrapped__, "__length_hint__"):
            if "__length_hint__" not in class_attrs:
                cls.__length_hint__ = __wrapper_length_hint__
        elif getattr(cls, "__length_hint__", None) is __wrapper_length_hint__:
            delattr(cls, "__length_hint__")

        if hasattr(self.__wrapped__, "__await__"):
            if "__await__" not in class_attrs:
                cls.__await__ = __wrapper_await__
        elif getattr(cls, "__await__", None) is __wrapper_await__:
            delattr(cls, "__await__")

        if hasattr(self.__wrapped__, "__get__"):
            if "__get__" not in class_attrs:
                cls.__get__ = __wrapper_get__
        elif getattr(cls, "__get__", None) is __wrapper_get__:
            delattr(cls, "__get__")

        if hasattr(self.__wrapped__, "__set__"):
            if "__set__" not in class_attrs:
                cls.__set__ = __wrapper_set__
        elif getattr(cls, "__set__", None) is __wrapper_set__:
            delattr(cls, "__set__")

        if hasattr(self.__wrapped__, "__delete__"):
            if "__delete__" not in class_attrs:
                cls.__delete__ = __wrapper_delete__
        elif getattr(cls, "__delete__", None) is __wrapper_delete__:
            delattr(cls, "__delete__")

        if hasattr(self.__wrapped__, "__set_name__"):
            if "__set_name__" not in class_attrs:
                cls.__set_name__ = __wrapper_set_name__
        elif getattr(cls, "__set_name__", None) is __wrapper_set_name__:
            delattr(cls, "__set_name__")


class LazyObjectProxy(AutoObjectProxy):
    """An object proxy which can generate/create the wrapped object on demand
    when it is first needed.
    """

    def __new__(cls, callback=None):
        return super().__new__(cls, None)

    def __init__(self, callback=None):
        """Initialize the object proxy with wrapped object as `None` but due
        to presence of special `__wrapped_factory__` attribute addded first,
        this will actually trigger the deferred creation of the wrapped object
        when first needed.
        """

        if callback is not None:
            self.__wrapped_factory__ = callback

        super().__init__(None)

    __wrapped_initialized__ = False

    def __wrapped_factory__(self):
        return None

    def __wrapped_get__(self):
        """Gets the wrapped object, creating it if necessary."""

        # We synchronize on the class type, which will be unique to this instance
        # since we inherit from `AutoObjectProxy` which creates a new class
        # for each instance. If we synchronize on `self` or the method then
        # we can end up in infinite recursion via `__getattr__()`.

        with synchronized(type(self)):
            # We were called because `__wrapped__` was not set, but because of
            # multiple threads we may find that it has been set by the time
            # we get the lock. So check again now whether `__wrapped__` is set.
            # If it is then just return it, otherwise call the factory to
            # create it.

            if self.__wrapped_initialized__:
                return self.__wrapped__

            self.__wrapped__ = self.__wrapped_factory__()

            self.__wrapped_initialized__ = True

            return self.__wrapped__


def lazy_import(name, attribute=None):
    """Lazily imports the module `name`, returning a `LazyObjectProxy` which
    will import the module when it is first needed. When `name is a dotted name,
    then the full dotted name is imported and the last module is taken as the
    target. If `attribute` is provided then it is used to retrieve an attribute
    from the module.
    """

    def _import():
        module = __import__(name, fromlist=[""])

        if attribute is not None:
            return getattr(module, attribute)

        return module

    return LazyObjectProxy(_import)
