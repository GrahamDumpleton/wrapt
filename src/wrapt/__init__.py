"""
Wrapt is a library for decorators, wrappers and monkey patching.
"""

__version_info__ = ("2", "0", "0", "rc2")
__version__ = ".".join(__version_info__)

from .__wrapt__ import (
    BoundFunctionWrapper,
    CallableObjectProxy,
    FunctionWrapper,
    ObjectProxy,
    PartialCallableObjectProxy,
    partial,
)
from .decorators import AdapterFactory, adapter_factory, decorator, synchronized
from .importer import (
    discover_post_import_hooks,
    notify_module_loaded,
    register_post_import_hook,
    when_imported,
)
from .patches import (
    apply_patch,
    function_wrapper,
    patch_function_wrapper,
    resolve_path,
    transient_function_wrapper,
    wrap_function_wrapper,
    wrap_object,
    wrap_object_attribute,
)
from .weakrefs import WeakFunctionProxy

__all__ = (
    "BoundFunctionWrapper",
    "CallableObjectProxy",
    "FunctionWrapper",
    "ObjectProxy",
    "PartialCallableObjectProxy",
    "partial",
    "AdapterFactory",
    "adapter_factory",
    "decorator",
    "synchronized",
    "discover_post_import_hooks",
    "notify_module_loaded",
    "register_post_import_hook",
    "when_imported",
    "apply_patch",
    "function_wrapper",
    "patch_function_wrapper",
    "resolve_path",
    "transient_function_wrapper",
    "wrap_function_wrapper",
    "wrap_object",
    "wrap_object_attribute",
    "WeakFunctionProxy",
)
