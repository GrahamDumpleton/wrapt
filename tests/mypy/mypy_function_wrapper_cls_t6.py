"""
This example demonstrates the incorrect usage of the FunctionWrapper type.

It covers the following cases:
- Wrapper as non callable object
- Wrapper as None object

Should fail mypy type checking for incorrect cases.
"""

from wrapt import FunctionWrapper


def function(x: int, y: str = "default") -> str:
    """A simple function to be wrapped."""
    return f"{x}: {y}"


wrapped_function = FunctionWrapper(function, "string")
wrapped_function = FunctionWrapper(function, None)
