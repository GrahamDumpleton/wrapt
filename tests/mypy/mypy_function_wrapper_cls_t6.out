tests/mypy/mypy_function_wrapper_cls_t6.py:19: error: Argument 2 to "FunctionWrapper" has incompatible type "str"; expected "Callable[[Callable[[int, str], str], Any, tuple[Any, ...], dict[str, Any]], str] | Callable[[type[Any], Callable[[int, str], str], Any, tuple[Any, ...], dict[str, Any]], str] | Callable[[Any, Callable[[int, str], str], Any, tuple[Any, ...], dict[str, Any]], str]"  [arg-type]
tests/mypy/mypy_function_wrapper_cls_t6.py:20: error: Argument 2 to "FunctionWrapper" has incompatible type "None"; expected "Callable[[Callable[[int, str], str], Any, tuple[Any, ...], dict[str, Any]], str] | Callable[[type[Any], Callable[[int, str], str], Any, tuple[Any, ...], dict[str, Any]], str] | Callable[[Any, Callable[[int, str], str], Any, tuple[Any, ...], dict[str, Any]], str]"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
